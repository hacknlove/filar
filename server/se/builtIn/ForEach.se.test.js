const { DOMParser } = require("linkedom");
const { processAllElements } = require("../../tree/processAllElements");
const { ServerElementsMap } = require("../common");
const parser = new DOMParser();

const ForEach = require("./ForEach.se");

describe("ForEach", () => {
  it("throws if the iterator is empty", async () => {
    jest.spyOn(console, "error").mockImplementation(() => {});
    const document = parser.parseFromString(`
                <div Title="'My title'">
                  <ul>
                    <ForEach>
                      <li>{{Item}}</li>
                    </ForEach>
                  </ul>
                </div>
            `);

    ServerElementsMap.set("ForEach", ForEach);

    await expect(() => processAllElements(document)).rejects.toThrow();
  });

  it("renders the children for each item in the iterator", async () => {
    const document = parser.parseFromString(`
        <div Title="'My title'">
          <ul>
            <ForEach key="Item" of="['one', 'two', 'three']">
              <li>{{Item}}</li>
            </ForEach>
          </ul>
        </div>
    `);

    ServerElementsMap.set("ForEach", ForEach);

    await processAllElements(document, {
      filePath: "test/index.html",
      __islands: {},
      __ce: {},
    });

    expect(document.toString()).toMatchSnapshot();
  });

  it("filters out with the filter", async () => {
    const document = parser.parseFromString(`
        <div Title="'My title'">
          <ul>
            <ForEach key="Item" of="['one', 'two', 'three', 'four', 'five', 'six']" filter="Item.length > 3">
              <li>{{Item}}</li>
            </ForEach>
          </ul>
        </div>
    `);

    ServerElementsMap.set("ForEach", ForEach);

    await processAllElements(document, {
      filePath: "test/index.html",
      __islands: {},
      __ce: {},
    });

    expect(document.toString()).toMatchSnapshot();
  });
});
