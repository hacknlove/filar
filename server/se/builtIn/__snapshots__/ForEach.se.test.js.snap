// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForEach filters out with the filter 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
        <div>
          <ul>
            
              <li>three</li>
            
              <li>four</li>
            
              <li>five</li>
            
          </ul>
        </div>
    "
`;

exports[`ForEach renders the children for each item in the iterator 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
        <div>
          <ul>
            
              <li>one</li>
            
              <li>two</li>
            
              <li>three</li>
            
          </ul>
        </div>
    "
`;
